delivery_procurement_group_carrier/models/stock_move.py

```diff
@@ -8,6 +8,13 @@ class StockMove(models.Model):
     def _get_new_picking_values(self):
         values = super()._get_new_picking_values()
         carrier = self.group_id.carrier_id
-        if carrier and self.picking_type_id.code in ["outgoing", "incoming"]:
+        if carrier and (
+            self.picking_type_id.code in ["outgoing", "incoming"]
+            or any(rule.propagate_carrier for rule in self.rule_id)
+        ):
             values["carrier_id"] = carrier.id
         return values
+
+    def _key_assign_picking(self):
+        keys = super()._key_assign_picking()
+        return keys + (self.group_id.carrier_id,)
```

Comment:
Why is needed? The group is already in the keys.
https://github.com/odoo/odoo/blob/52aa6231130ea165fdb44e6370ec3e396b7603cc/addons/stock/models/stock_move.py#L1346